### ISSUE TITLE
Forecasting with Neural Network - Sales History input recomendation

### ISSUE THREAD

**Issue by swaheed2** [cazala/synaptic#183]
**Training**

**Product ID &nbsp;&nbsp;&nbsp;&nbsp;YEAR&nbsp;&nbsp;&nbsp;&nbsp;MONTH&nbsp;&nbsp;&nbsp;&nbsp;QUANTITY (Output)**

abc-X-6 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  2012  &nbsp;&nbsp;&nbsp; &nbsp;   1     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;        2.00

wer-X-6 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  2013  &nbsp;&nbsp;&nbsp; &nbsp;   1     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;        2.00

abc-X-6 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  2012  &nbsp;&nbsp;&nbsp; &nbsp;   1     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;        2.00

bvd-X-6 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  2014  &nbsp;&nbsp;&nbsp; &nbsp;   1     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;        2.00

sdfv-X-6 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  2015  &nbsp;&nbsp;&nbsp; &nbsp;   12     &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;        2.00

**Activate Example:**
bvd-X-6   &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;    2016   &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1             &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;    ?

There are about **2000** rows of data, with about **150 product ids**.

If can normalize the data as explained here https://github.com/cazala/synaptic/wiki/Normalization-101

So normalizing ```productId``` would make a bit array of 150 length like: ```[0,0,...1]```

Year from 2012 to 2015 would be a like:
```
 [0,0,0,1] 
 [0,0,1,0] 
 [0,1,0,0] 
 [1,0,0,0] 
```

This would make the input array to be about ~165 in length, and output 1. 

I don't have any experience with neural networks. Can you recommend if using Neural Networks to forecast next month quantity is a good use case? 

I only want to forecast next month, not multiple months in the future.

**Comment by pauljohnleonard** [comment]
Yes NN are good for this type of problem.   You train the network sliding a feedforward network in time using historic data.  Your inputs can be the recent history (say last 10 months).   Your output is the next month. You could also try a RNN which can remember history.

**Comment by Jabher** [comment]
I'd rather say you need some embedding here, not one-hot encoding. Also you will probably need some feature-engineering here.  Also you will probably need to merge year+month (in some kind of absolute time, like 2015 * 12 + 2 for feb 2015) and use month as additional feature for processing the seasonal factors.
 
Also it is historical data, so, well, it's always a complicated stuff. This quora post is nice for that: https://www.quora.com/What-would-be-the-best-inputs-for-a-Neural-Network-algorithm-trying-to-predict-the-stock-market

Another thing is that 2000 rows and 150 product ids is too small dataset to predict something. If I were you I would probably try to create a linear regression for sliding window of each item, for item relative to all sales, and for overall sales (maybe target should be log-ed or non-linearly transformed in some other way, this is totally on you). After that, if result is still not satisfying, you should probably look for something more complicated.
